custom feedback

f0rki opened the pull request:
starting a PR now, since we might want to start merging this :) 

* [x] a bit of refactoring to allow breakpoint handlers to set custom feedback
* [x] some new feedback types, such as max feedback (and also min feedback) for values.
    * some obvious usecases, such as minimizing by string prefix distance are included.
    * TODO: are there other interesting feedbacks that we can provide?
        * [x] implement [value profile](https://www.llvm.org/docs/LibFuzzer.html#value-profile) as custom feedback? i.e., a `popcount(~(a ^ b))` for comparisons.
* [x] use custom feedback in other commands
    * [x] `minimize` - minimize corpus entry by coverage/custom feedback
    * [x] ~~...?~~ non for now.
* [x] redqueen feature compatibility
    * ~~currently does not compile.~~
    * either rebase on the redqueen-rework branch or disable for now and integrate later.
* [x] add example for using custom feedback (e.g., klee maze code)
* [x] add a `test.sh` similar to the other examples
* [x] add missing doc comments
* [x] ~~do we need the tagged feedback mechanism?~~ no. removed.
* [x] rebase on latest dev.

Commit:
progress on custom feedback

Commit:
Added a `snapchange::prelude` pub mod to quickly get access to everything needed to get started.
Pub use of the rand and anyhow crates - those are often needed by users of snapchange.

Commit:
Fixed `record_min_prefix_dist` function.

Commit:
Implemented convenience function to record a pair of u32 together.

f0rki commented on src/feedback.rs:
```diff
@@ -0,0 +1,235 @@
+
+    #[cfg(feature = "redqueen")]
+    #[serde(skip)]
+    Redqueen((VirtAddr, RFlags)),
```
@corydu I copied this from the old redqueen code, but this actually doesn't compile because `RFlags` is not `Serializable`. Not sure if it is really needed though.

corydu replied:
Yeah we could just use a u64 for this and not the actual `RFlags` itself.

Commit:
Fixed redqueen related code such that it compiles.

Commit:
added example to showcase custom feedback

Commit:
Improved custom feedback code.

* Added documentation in the feedback module. Removed the
  `#[allow(missing_docs)]`
* Fixed all doc tests in `feedback.rs`.
* Made hitcount code coverage the default in the `FeedbackTracker`.
  Coverage breakpoints are scratched away in the snapshot once they have
  been hit, so the normal fuzzer will not see any difference, since a
  coverage breakpoint is not recorded twice.
  However, we can use custom breakpoint hooks to record code coverage
  hitcounts for selective addresses or utilize hitcounts in other parts
  of snapchange (e.g., minimize or redqueen).
* Added a helper for value profile custom feedback.
* Removed the tagged feedback. Most usescases should be covered well enough
  by using hashing, i.e., `feedback.record_tagged(t, v)` becomes
  `feedback.record_hashed([t, v])`. While this is prone to hash
  collisions, I don't think it is too bad. If we find this is a real
  problem we can add the tagged feedback again later.
* Switched from the older fxhash crate to the "official" rustc-hash
  crate (same algorithm though).
* Moved all feedback to hashmap/sets.

Commit:
Updated Example 6

* Fixed Typo (lose vs. loose)
* Code Formatting
* implemented `minimize` for `WasdArray`
* added a `test.sh` similar to the other examples
* switched to the `record_pair` custom feedback api.
* Added a tutorial.

Commit:
`minimize` inputs according to feedback (code coverage or custom).

* Reworked the `minimize` code, fixing some legacy copy-paste issues.
* introduced `FuzzVm::gather_feedback` to gather oneshot feedback about
  the execution of a single input.
* Updated the `FeedbackTracker` with several convenience methods for
  comparing two instances.
  * e.g., different classification of hitcounts into buckets, etc.
  * otherwise minimization is often not as useful.
* Updated the `minimize` cli with several flags to control when
  minimization is successful
  * `--ignore-stack` and `--ignore-console-output` to disable comparing
    these.
  * `--consider-coverage none|symbols|basic-block|hitcounts` to
    configure how coverage is taken into account.

Commit:
Fix small typo in documentation.

Commit:
misc updates (formatting, typos, documentation)

Commit:
fixed resetting wrong maze target in test.sh / not setting the snapshot project dir (silently errored..)
